File: CodeTester-IntelliJ-IDEA/src/main/java/de/fxnm/runnable/BaseRunnable.java
package de.fxnm.runnable;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import de.fxnm.listener.FeedbackListener;

public abstract class BaseRunnable<T> implements Runnable {

    public static Logger LOG;
    private final List<FeedbackListener> listeners = new LinkedList<>();
    private final Project project;
    private Boolean finished = false;

    public BaseRunnable(final Project project, final Class<T> loggerClass) {
        this.project = project;
        LOG = Logger.getInstance(loggerClass);
    }

    public Project project() {
        return this.project;
    }

    public void addListener(final FeedbackListener listener) {
        this.listeners.add(listener);
    }

    public void startRunnable(final String processName, final Object... details) {
        LOG.info("Started Callable, " + Arrays.deepToString(details));
        this.backgroundLoadingBar(processName);
        this.listeners.forEach(listener -> listener.scanStarting(details));
    }

    public void finishedRunnable(final Object... details) {
        if (this.finished.equals(Boolean.TRUE)) {
            LOG.info("Finished or failed Callable got already executed");
            return;
        }

        LOG.info("Finished Callable successful");
        this.finished = true;
        this.listeners.forEach(listener -> listener.scanCompleted(details));
    }

    public void failedRunnable(final Object... details) {
        if (this.finished.equals(Boolean.TRUE)) {
            LOG.info("Finished or failed Callable got already executed, " + Arrays.deepToString(details));
            return;
        }

        LOG.info("Finished Callable with error, " + Arrays.deepToString(details));
        this.finished = true;
        this.listeners.forEach(listener -> listener.scanFailed(details));
    }

    private void backgroundLoadingBar(final String processName) {
        ProgressManager.getInstance().run(new Task.Backgroundable(this.project(), processName) {
            public void run(@NotNull final ProgressIndicator progressIndicator) {
                while (BaseRunnable.this.finished.equals(Boolean.FALSE)) {
                    progressIndicator.setIndeterminate(true);
                }
            }
        });
    }
}

File: CodeTester-IntelliJ-IDEA/src/main/java/de/fxnm/runnable/BaseRunnable.kt
package de.fxnm.runnable

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task.Backgroundable
import com.intellij.openapi.project.Project
import de.fxnm.listener.Listener
import de.fxnm.util.CodeTesterBundle
import java.util.LinkedList
import java.util.function.Consumer

abstract class BaseRunnable(private val project: Project, loggerClass: Class<*>) : Runnable {
    private val listeners: MutableList<Listener> = LinkedList()
    private var finished = false
    private var LOG: Logger = Logger.getInstance(loggerClass)

    fun project(): Project {
        return project
    }

    fun addListener(listener: Listener) {
        listeners.add(listener)
    }

    @JvmOverloads
    fun startRunnable(
        loggerMessage: String,
        toolWindowMessage: String,
        backGroundProcessName: String,
        argumentOne: Any? = null,
        argumentTwo: Any? = null,
        argumentThree: Any? = null
    ) {
        LOG.info("$loggerMessage - Arguments: ${argumentOne.toString()} | ${argumentTwo.toString()} | ${argumentThree.toString()} - Background ProcessName: $backGroundProcessName")

        backgroundLoadingBar(backGroundProcessName)

        listeners.forEach(Consumer { listener: Listener ->
            listener.scanStarting(
                toolWindowMessage, backGroundProcessName, argumentOne, argumentTwo, argumentThree
            )
        })
    }

    @JvmOverloads
    fun finishedRunnable(
        loggerMessage: String,
        toolWindowMessage: String,
        argumentOne: Any? = null,
        argumentTwo: Any? = null,
        argumentThree: Any? = null
    ) {
        if (finished) {
            LOG.info(CodeTesterBundle.message("plugin.runnable.baseRunnable.alreadyFinished"))
            return
        }

        LOG.info("$loggerMessage - Arguments: $argumentOne | ${argumentTwo.toString()} | ${argumentThree.toString()}")

        finished = true
        listeners.forEach(Consumer { listener: Listener ->
            listener.scanCompleted(
                toolWindowMessage, argumentOne, argumentTwo, argumentThree
            )
        })
    }

    @JvmOverloads
    fun failedRunnable(
        loggerMessage: String,
        toolWindowMessage: String,
        throwable: Throwable? = null,
        argumentOne: Any? = null,
        argumentTwo: Any? = null,
        argumentThree: Any? = null
    ) {
        if (finished) {
            LOG.info(CodeTesterBundle.message("plugin.runnable.baseRunnable.alreadyFinished"))
            return
        }

        LOG.info(
            "$loggerMessage - Arguments: ${argumentOne.toString()} | ${argumentTwo.toString()} | ${argumentThree.toString()}",
            throwable
        )

        finished = true
        listeners.forEach(Consumer { listener: Listener ->
            listener.scanFailed(
                toolWindowMessage, throwable, argumentOne, argumentTwo, argumentThree
            )
        })
    }

    private fun backgroundLoadingBar(processName: String) {
        if (ApplicationManager.getApplication().isUnitTestMode) {
            return
        }
        ProgressManager.getInstance().run(object : Backgroundable(project(), processName, false) {
            override fun run(progressIndicator: ProgressIndicator) {
                LOG.info(
                    String.format(
                        CodeTesterBundle.message(
                            "plugin.runnable.baseRunnable.backgroundLoadingBar.start"
                        ), processName
                    )
                )
                while (!finished) {
                    progressIndicator.isIndeterminate = true
                }
                LOG.info(
                    String.format(
                        CodeTesterBundle.message(
                            "plugin.runnable.baseRunnable.backgroundLoadingBar.finished"
                        ), processName
                    )
                )
            }
        })
    }
}